specify/specify7
# MOTs business logic: Clear CO form when CO Type changes

Fixes #5088 

Using business rules to clear CO values when COType changes. Needs #5098 to be fixed to test with configuring a pick list in Schema Config. Until then, it can be tested with CoType as a QCBX.

The following fields are ignored when clearing for better UX or based on backend restrictions (for eg: backend needs a non-null value for version)
https://github.com/specify/specify7/blob/64fd7ea4643fa080f9adf41a1f7575de7ca7f52d/specifyweb/frontend/js_src/lib/components/DataModel/businessRuleDefs.ts#L155-L161


### Checklist

      and self-explanatory (or properly documented)

### Testing instructions




#### Set up a QCBX for CollectionObjectType (like in #5127):
Ideally, COType would be set as a picklist. However, until #5098 is fixed, it is not possible to configure one for COType. Instead for this PR, COType can be set up as a QCBX. 
- Go to App Resources > Type Searches (found under Global Resources)
- Add the following line in the XML editor:
```xml
<typesearch tableid="1015"  name="CollectionObjectType" searchfield="name" displaycols="name" format="%s" uifieldformatter="" dataobjformatter="" system="true"/>
```
- Click on the Form Definition tab
- Select the correct `.views` file for the collection you are in
- Go the the Collection Object form definition
- Add the following lines inside an existing `<row>`:
```xml
<cell type="label" labelfor="111"/>
<cell type="field" id="111" name="collectionObjectType" uitype="querycbx" />
```
Your CO form should look something like this
![image](https://github.com/user-attachments/assets/a4c9bb41-f0d5-48c2-b4d2-0d297dd4927c)

#### Testing the PR
- Go to Data Entry -> Collection Object
- Enter some determinations
- Enter data in different basic fields (text/number fields)
- Change CollectionObjectType value
- [ ] Verify all simple fields are cleared
- [ ] Verify any added determinations are deleted
- Re-run the test for an already existing Collection Object
